### ISSUE TITLE
XOAUTH2 SASL client

### ISSUE THREAD

**Issue by k-wall** [vcabbage/amqp#190]
My use-case requires requires [XOAUTH2](https://developers.google.com/gmail/imap/xoauth2-protocol#the_sasl_xoauth2_mechanism) SASL authentication with a bearer token.  vcabbage/amqp does not currently support this mechanism.  I am offering to raise a PR with support for it.

I am proposing to add `func ConnSASLXOAUTH2(username, bearer string) ConnOption` and supporting tests.

There is a problem that arises when using XOAUTH2 over AMQP SASL that I must also handle.  The maximum SASL frame size defined by AMQP 1.0 is limited to 512 bytes.  This is often too small to handle bearer tokens.  Unfortunately, the protocol defines no way to raise the *SASL* frame limit or chunk the SASL payload. This was discussed on the Apache Qpid mailing list [1].

To solve this, I also need to add `func PeerMaxFrameSize(int32) ConnOption` which may be used if it is known a priori that such a AMQP SASL frame will be acceptable to the server.

Does this sound like a reasonable enhancement?

[1] http://qpid.2158936.n2.nabble.com/Re-SASL-GSSAPI-and-MIN-MAX-FRAME-SIZE-td7665117.html

(corrected to Apache Qpid discussion on SASL max frame size)

**Comment by vcabbage** [comment]
Overall, sounds good to me.

My brief scan of the documentation suggests that this can be implemented without any external dependencies, is that correct?

I think the wrong link was added for the Apache Qpid mailing list reference. Presumably the `PeerMaxFrameSize` is required since the default is 512 during the SASL phase before Open has been exchanged? Instead of the additional `ConnOption` would it make sense to have `ConnSASLXOATH2` make this adjustment? `PeerMaxFrameSize` might give the incorrect impression that the client can set or influence the server's max frame size.

**Comment by k-wall** [comment]
> Overall, sounds good to me.
> 
> My brief scan of the documentation suggests that this can be implemented without any external dependencies, is that correct?

Yes, that's my expectation.  No new dependencies.

> 
> I think the wrong link was added for the Apache Qpid mailing list reference.

Sorry, I muddled the links.   The correct link is:

http://qpid.2158936.n2.nabble.com/Re-SASL-GSSAPI-and-MIN-MAX-FRAME-SIZE-td7665117.html

I edited my original issue comment above.

> Presumably the `PeerMaxFrameSize` is required since the default is 512 during the SASL phase before Open has been exchanged?

Exactly.

> Instead of the additional `ConnOption` would it make sense to have `ConnSASLXOATH2` make this adjustment? `PeerMaxFrameSize` might give the incorrect impression that the client can set or influence the server's max frame size.

Having ConnSASLXOATH2 make the adjustment is possible, and having fewer options to set is always appealing.   However, the I worry about opening an attack vector.  If an attacker were able to  influence the size of the username or bearer being passed to ConnSASLXOAUTH2 func, they could then abuse vcabbage AMQP to generate excessively sized SASL frames which could DoS/crash an AMQP server.   Having a separate control at least gives the user of the vcabbage AMQP library the means to control the upper limit.

How about `ConnSASLXOAUTH2(username, bearer string, saslMaxFrameSize int32)`?  This avoids the misleading impressions that a standalone PeerMaxFrameSize would give whilst retaining the explicit control that the user would have to tune for their OAUTH2 provider's token size.
